osu.Game/Screens/Edit/Editor.cs

```diff
@@ -135,6 +139,19 @@ private void load(OsuColour colours, OsuConfigManager config)
             {
                 playableBeatmap = loadableBeatmap.GetPlayableBeatmap(loadableBeatmap.BeatmapInfo.Ruleset);
 
+                // the game-wide beatmap doesn't necessarily have all of the difficulties in its BeatmapSet.Beatmaps property, due to EF idiosyncrasies.
+                // re-query explicitly during this process to prevent difficulties getting truncated on beatmap save.
+
+                // it's important to query the beatmap *set*, rather than the single beatmap, to ensure all beatmaps from the set are present,
+                // and to replace the playableBeatmap.BeatmapInfo reference with the one re-fetched from database to ensure changes to the editor beatmap's BeatmapInfo are persisted to EF.
+                var refetchedBeatmapSet = beatmapManager.QueryBeatmapSet(bs => bs.ID == Beatmap.Value.BeatmapSetInfo.ID);
+
+                if (refetchedBeatmapSet != null)
+                {
+                    playableBeatmap.BeatmapInfo = refetchedBeatmapSet.Beatmaps.Single(b => b.ID == playableBeatmap.BeatmapInfo.ID);
+                    playableBeatmap.BeatmapInfo.Metadata ??= refetchedBeatmapSet.Metadata;
```

Comment:
Note that this is now assigning to `BeatmapInfo` rather than just `BeatmapInfo.BeatmapSet` as the old code used to do, as per the comment above. As it turns out those two references could differ before, which is why the difficulty name changes didn't save.

As for the metadata shuffling, I needed to do this to prevent crashes in `LegacyBeatmapEncoder`, which would sometimes access `EditorBeatmap.BeatmapInfo.Metadata` (e.g. when trying to encode path to the background file) which isn't guaranteed to be present. This is a carbon copy of similar workarounds in `BeatmapManager.GetWorkingBeatmap()` and `BeatmapCarousel.createCarouselSet()`.